src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts

```diff
@@ -0,0 +1,195 @@
+import { toAbsoluteUri } from '../Services/NavigationManager';
+
+export class WebAssemblyResourceLoader {
+  private usedCacheKeys: { [key: string]: boolean } = {};
+  private networkLoads: { [name: string]: LoadLogEntry } = {};
+  private cacheLoads: { [name: string]: LoadLogEntry } = {};
+
+  static async initAsync(): Promise<WebAssemblyResourceLoader> {
+    const bootConfigResponse = await fetch('_framework/blazor.boot.json', {
+      method: 'GET',
+      credentials: 'include'
+    });
+
+    const relativeBaseHref = document.baseURI.substring(document.location.origin.length);
+    const cacheName = `blazor-resources-${relativeBaseHref}`;
+    return new WebAssemblyResourceLoader(
+      await bootConfigResponse.json(),
+      await caches.open(cacheName));
+  }
+
+  constructor (public readonly bootConfig: BootJsonData, private cache: Cache)
+  {
+  }
+
+  loadResources(resources: ResourceList, url: (name: string) => string): LoadingResource[] {
+    return Object.keys(resources)
+      .map(name => this.loadResource(name, url(name), resources[name]));
+  }
+
+  loadResource(name: string, url: string, contentHash: string): LoadingResource {
+    const networkFetchOptions: RequestInit = { cache: 'no-cache' };
+    if (!this.bootConfig.cacheBootResources) {
+      // Bypass the entire cache flow, including checking hashes, etc.
+      // This gives developers an easy opt-out if they don't like anything about the default cache mechanism
+      const response = fetch(url, networkFetchOptions);
+      const data = response.then(r => r.clone().arrayBuffer());
+      return { name, url, response, data };
+    }
+
+    if (!contentHash || contentHash.length === 0) {
+      throw new Error('Content hash is required');
+    }
+
+    const cacheKey = toAbsoluteUri(`${url}.${contentHash}`);
+    this.usedCacheKeys[cacheKey] = true;
+
+    const responseInfoPromise = (async () => {
+      // Try to load from cache
+      const cachedResponse = await this.cache.match(cacheKey);
+      if (cachedResponse) {
+        const responseBytes = parseInt(cachedResponse.headers.get('content-length') || '0');
+        this.cacheLoads[name] = { responseBytes };
+        return { response: cachedResponse, isNetworkResponse: false };
+      }
+
+      // It's not cached (or caching is now disabled), so fetch from network
+      const networkResponse = await fetch(url, networkFetchOptions);
+      return { response: networkResponse, isNetworkResponse: true };
+    })();
```

Comment:
I did it like this on purpose because it lets the compiler use an anonymously inferred type for the results. TBH I feel it's a pretty nice way of starting a background `Promise<T>` where `T` is anonymous within the flow of the outer function.

If you feel strongly I will move it outside, but it will be quite a bit more verbose and not as obvious that the logic is only intended for use in this one place.

Reply:
I got lost reading the code, and went back and forth a few times. It might just be me, if all we are saving is a typing, I think I would err on the side of clarity, but I leave it up to you.